core/src/message/signal.rs

```diff
@@ -0,0 +1,82 @@
+// This file is part of Gear.
+
+// Copyright (C) 2022 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{
+    ids::{MessageId, ProgramId},
+    message::{Dispatch, DispatchKind, ExitCode, Message, Payload, ReplyDetails},
+};
+use codec::{Decode, Encode};
+use scale_info::TypeInfo;
+
+/// Message for Reply entry point.
+/// [`ReplyMessage`] is unique because of storing [`MessageId`] from message on what it replies, and can be the only one per some message execution.
+#[derive(Clone, Default, Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Decode, Encode, TypeInfo)]
+pub struct SignalMessage {
+    /// Message id.
+    id: MessageId,
+    /// Message payload.
+    payload: Payload,
```

Comment:
I see no reason in payload here.

Signal will have just message id it associates with and exit code

Reply:
How we supposed to send trap reason to user then?

Reply:
> How we supposed to send trap reason to user then?

For me seems reasonable that "signal" identifies itself via just code (exit_code)

Reply:
And reply_to identifies which message produced this signal